codespan-reporting/src/lib.rs

```diff
@@ -1,4 +1,285 @@
+    pub source: Source,
+}
+
+impl<Source> Line<Source>
+where
+    Source: AsRef<str>,
+{
+    /// The column index at the given byte index in the source file.
+    /// This is the number of characters to the given byte index.
+    ///
+    /// If the byte index is smaller than the start of the line, then `0` is returned.
+    /// If the byte index is past the end of the line, the column index of the last
+    /// character `+ 1` is returned.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use codespan_reporting::Line;
+    ///
+    /// let line = Line {
+    ///     start: 2,
+    ///     number: 2,
+    ///     source: "🗻∈🌏",
+    /// };
+    ///
+    /// assert_eq!(line.column_index(0), 0);
+    /// assert_eq!(line.column_index(line.start + 0), 0);
+    /// // FIXME: Snap column index to the previous character boundary
+    /// // assert_eq!(line.column_index(line.start + 1), 0);
+    /// assert_eq!(line.column_index(line.start + 4), 1);
+    /// assert_eq!(line.column_index(line.start + 8), 3);
+    /// assert_eq!(line.column_index(line.start + 48), 3);
+    /// ```
+    pub fn column_index(&self, byte_index: usize) -> usize {
+        match byte_index.checked_sub(self.start) {
+            None => 0,
+            Some(relative_index) => {
+                let line_source = self.source.as_ref();
+
+                line_source
+                    .char_indices()
+                    .map(|(i, _)| i)
+                    .take_while(|i| *i < relative_index)
+                    .count()
+            },
+        }
+    }
+
+    /// The 1-indexed column number at the given byte index.
+    pub fn column_number(&self, byte_index: usize) -> usize {
+        self.column_index(byte_index) + 1
+    }
+}
+
+/// Files that can be used for pretty printing.
+pub trait Files {
+    type FileId: Copy + PartialEq + PartialOrd + Eq + Ord + std::hash::Hash;
+    type Origin: std::fmt::Display;
+    type LineSource: AsRef<str>;
+
+    /// The origin of a file.
+    fn origin(&self, id: Self::FileId) -> Option<Self::Origin>;
+
+    /// The line at the given index.
+    fn line(&self, id: Self::FileId, line_index: usize) -> Option<Line<Self::LineSource>>;
+
+    /// The index of the line at the given byte index.
+    fn line_index(&self, id: Self::FileId, byte_index: usize) -> Option<usize>;
+
+    /// The location of the given byte index.
+    fn location(&self, id: Self::FileId, byte_index: usize) -> Option<Location> {
+        let line_index = self.line_index(id, byte_index)?;
+        let line = self.line(id, line_index)?;
+
+        Some(Location {
+            line_number: line.number,
+            column_number: line.column_number(byte_index),
+        })
+    }
+}
+
+/// A single source file.
+///
+/// This is useful for simple language tests, but it might be worth creating a
+/// custom implementation when a language scales beyond a certain size.
+#[derive(Debug, Clone)]
+pub struct SimpleFile<Origin, Source> {
```

Comment:
Do you plan to add more implementations of `Files` in the future? If so it might be worth making this a submodule so `lib.rs` doesn't get cluttered.

Reply:
Hmm I could do that!

Reply:
I was thinking of re-exporting the diagnostic and files stuff from the top level FWIW, but I might do that in a separate PR.